### ISSUE TITLE
``__getitem__`` tuple error `call-non-callable`

### ISSUE THREAD

**Issue by AA-Turner** [astral-sh/ty#468]
### Summary

I don't believe this is a duplicate, apologies if otherwise.

```python
from typing import Any
class Spam:
    def __getitem__(self, item: int | slice) -> Any:
        return (1, 'spam', None)[item]
```

`ty` reports `Method `__getitem__` of type `Overload[(key: SupportsIndex, /) -> Unknown, (key: slice[Any, Any, Any], /) -> tuple[Unknown, ...]]` is not callable on object of type `tuple[Literal[1], Literal["spam"], None]` (call-non-callable) [Ln 4, Col 16]`

A

https://play.ty.dev/13c25386-6806-4686-880c-7891a3fb9044

### Version

play.ty.dev reports 0.0.0, but locally `ty 0.0.1-alpha.5 (3b726d87a 2025-05-17)`

**Comment by sharkdp** [comment]
Thank you for reporting this!

It looks like we do not properly handle unions in overload resolution. Here's a self-contained example:
```py
from typing import overload

class A: ...
class B: ...

@overload
def f(x: A) -> int: return 1
@overload
def f(x: B) -> str: return "a"
def f(x: A | B) -> int | str:
    raise NotImplementedError

def _(x: A | B):
    reveal_type(f(x))  # should be int | str, ty: no-matching-overload
```
(https://play.ty.dev/359d3749-dab4-4857-9bb6-b127aaae0910)

**Comment by sharkdp** [comment]
Handling this properly will require "forking" the argument-type-checking algorithm in overload resolution for every possible combination of union elements. For a call like `f(x, y)` with `x: X1 | X2` and `y: Y1 | Y2`, we would need to check all combinations `(X1, Y1); (X1, Y2); (X2, Y1); (X2, Y2)`. If one of these combinations results in a call error, the entire call should fail. Otherwise, the resulting type will be a union of all return types. The important part is that each of these combinations can pick a different overload.

**Comment by MichaReiser** [comment]
CC: @dhruvmanila

**Comment by dhruvmanila** [comment]
Yes, this is exactly the thing that I'm currently working on :)

I'll convert this issue to be part of #104.
